### ISSUE TITLE
Be able to create a new project with prepoluated class names instead of boilerplate project names (MyApp, MyHomePage)

### ISSUE THREAD

**Issue by maryx** [flutter/flutter#13611]
This is just a feature request.

Using `flutter create myapp` and `File -> New Flutter Project...` in Android Studio/IntelliJ both result in the sample code for MyApp (the increment counter app) being created. Could we please have a way to:

 * instantiate a blank project?
 * or instantiate something without so much code already written (e.g. see below)?
 * or something that can customize the name of the app (i.e. if I named my Flutter Application hello_world, MyApp would be called HelloWorld?

Something like:
```dart
import 'package:flutter/material.dart';

void main() => runApp(new HelloWorld());

class HelloWorld extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new Container();
  }
}
```

Reasoning:
 * As a regular developer, the stuff in the sample app may not be relevant (classes would be in different files, comments would be removed). There's a bit of a mental overhead to renaming the classes. Chances are, you'd just delete all the code and start from scratch, so it would be great to have this option when creating a new project.
 * As a very new developer, one might be more interested in the just StatelessWidgets first (e.g. just how to do visual layouts), and worry about StatefulWidgets (the logic to the app) later.

**Comment by sethladd** [comment]
Eventually we'll need a way for the community to contribute templates and starter projects, because one size doesn't fit all, no matter how many options we offer via the tool.

Imagine a world where community members can contribute templates, and publish those templates to pub. Then imagine the flutter tool can ask pub "list me all the project templates" and offer them as options to the user. The user can then select a starter template, and the flutter tool installs the template package, and then creates an app from the template.

Overall, our goal should be to open up our system for community contributions and create an extensible world. Starter project templates are a good opportunity for that :)

**Comment by Hixie** [comment]
@maryx Can you elaborate on why this is needed? For a blank project, you can just create a new project, then select-all and press delete. Isn't that sufficient?

**Comment by maryx** [comment]
Yep, we could just delete the code. It'd be nice, though, to have the option to start with the blank project, which seems more like a "default", rather than the counter being the "default". For example, many web hosts give you a blank index.html and index.css page rather than a pre-made project. I'm not sure about mobile apps.

I get that most new projects are probably being started by people new to Flutter, so it's a useful reference to have the example app. When we get to the point in the future where most new projects are being started by people already with Flutter experience, it'd be nice to have a blank project option.

Ignoring the above, the most immediately useful feature would be to replace the class names with the name we type in in `flutter create <name>`. When following the [first app tutorial](https://flutter.io/getting-started/), I expected the boilerplate code to have named my classes, since the directory is already custom-named to <name>. E.g. if I type `flutter create weather`, a `weather` dir is created, and I would expect `new MyApp`, `class MyApp`, `class MyHomePage` to be `new WeatherApp`, `class WeatherApp`, and `class WeatherHomePage`.

**Comment by Hixie** [comment]
that wouldn't be black then right? that sounds like a separate feature request.

I think it makes sense to prepopulate the class names; i don't see much value in an actually blank main.dart given how easy it is to make it blank, especially since a black one would immediately report errors.

**Comment by maryx** [comment]
Sure, having a blank project can be a separate feature request.

**Comment by maryx** [comment]
I'll change the title of this one to be to have preopulated class names.

**Comment by devkabiir** [comment]
This looks like something I can do, in [flutter/packages/flutter_tools/templates/create/lib/main.dart.tmpl](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/create/lib/main.dart.tmpl) Replacing all instances of `My*` classes with `{{plguinDartClass}}` will expand to class names like `WeatherApp` & `WeatherAppHomePage`. But when an actual plugin is generated perhaps the class names should stay `MyApp` and `MyAppHomePage`. And maybe instead of using `{{pluginDartClass}}` there should be `{{packageDartClass}}` which might also prove useful in custom-templates for flutter apps.

> Imagine a world where community members can contribute templates, and publish those templates to pub. Then imagine the flutter tool can ask pub "list me all the project templates" and offer them as options to the user. The user can then select a starter template, and the flutter tool installs the template package, and then creates an app from the template. 

I've started Imagining 😄, I think the first step towards that would be to to allow setting a custom templates directory when creating a new app e.g. something like

```sh
flutter create myapp --local-template </path/to/local/template/directory>
# or
flutter create myapp --pub-template <pub_package_name>
```

I've started working on the `--local-template` part, and so far it seems to work. Idk how to implement the `--pub-template` part though, this would likely require a predefined structure for pub packages and a way for pub server to distinguish between template_packages and other packages so it can send correct options for user to choose from.

Now I want to contribute, Should I work on updating the default app template or continue working on `flutter_tools` package to allow custom-templates (#14719)? (I'm happy to do both too)

**Comment by bkonyi** [comment]
This is out of scope for the tool as it can be easily accomplished by third party packages (e.g., `package:mason` supports template based generation).
